aws/resource_aws_vpc_endpoint_subnet_association.go

```diff
@@ -46,10 +48,25 @@ func resourceAwsVpcEndpointSubnetAssociationCreate(d *schema.ResourceData, meta
 		return err
 	}
 
-	_, err = conn.ModifyVpcEndpoint(&ec2.ModifyVpcEndpointInput{
-		VpcEndpointId: aws.String(endpointId),
-		AddSubnetIds:  aws.StringSlice([]string{snId}),
-	})
+	// See https://github.com/terraform-providers/terraform-provider-aws/issues/3382.
+	// Prevent concurrent subnet association requests and delay between requests.
+	mk := "vpc_endpoint_subnet_association_" + endpointId
+	awsMutexKV.Lock(mk)
```

Comment:
I believe we prefer to implement solutions to this problem with [`resource.Retry()`](https://godoc.org/github.com/hashicorp/terraform/helper/resource#Retry) instead of a mutex for a few reasons:
* Handles cases where the concurrent modification might be happening outside Terraform
* Better handles EC2 eventual consistency as we do not want to guess about the potential delay (it might be shorter than a minute for the API call to work)
* Slightly simpler to understand logic (you just simply declare what error(s) you are retrying on with a timeout)

What do you think?

Reply:
I'll have to refresh my memory of what the underlying EC2 API's behavior is in this case.
I _think_ the behavior was that the `ModifyVpcEndpoint` API call just hangs waiting for a response in these situations.

Reply:
Okay, it'd definitely be good to check the behavior on that from the debug logs. If that is the case, where I'm guessing the SDK is automatically retrying when it probably shouldn't, we should probably fix that in by adding a handler in `config.go`. 😄

Reply:
When I manually perform the test steps in the AWS console I notice that the VPC Endpoint goes into a _pending_ state while the subnets are being associated. The code doesn't take that into consideration. I'm going to add a `WaitForState` call and see if that improves things.